integration-tests/pkg/mock_sensor/expect_conn.go

```diff
@@ -72,14 +72,90 @@ loop:
 			if conn.GetContainerId() != containerID {
 				continue loop
 			}
-
 			if len(s.Connections(containerID)) == n {
 				return s.Connections(containerID)
 			}
 		}
 	}
 }
 
+// checkIfConnectionsMatchExpected compares a list of expected and observed connection match exactly.
+func (s *MockSensor) checkIfConnectionsMatchExpected(t *testing.T, connections []types.NetworkInfo, expected []types.NetworkInfo) bool {
+	if len(connections) > len(expected) {
+		return assert.ElementsMatch(t, expected, connections, "networking connections do not match")
+	}
+
+	if len(connections) == len(expected) {
+		types.SortConnections(connections)
+		for i := range expected {
+			if !expected[i].Equal(connections[i]) {
+				return assert.ElementsMatch(t, expected, connections, "networking connections do not match")
+			}
+		}
+		return true
+	}
+	return false
+}
```

Comment:
Isn't this whole method the same as  just calling `assert.ElementsMatch(t, expected, connections, "networking connections do not match")`? From the testify docs:
```
ElementsMatch asserts that the specified listA(array, slice...) is equal to specified listB(array, slice...) ignoring the order of the elements. If there are duplicate elements, the number of appearances of each of them in both lists should match. 
```
https://pkg.go.dev/github.com/stretchr/testify/assert#ElementsMatch

Reply:
No. There is a difference. This uses the `Equal` method, which is not what `ElementsMatch` uses. The `Equal` method views timestamps as being equal if they are both non-nil or both nil, while `ElementsMatch` checks that the timestamps are actually the same. This is similar to what other existing parts of the code do. E.g. `ExpectEndpoints` checks if an endpoint is in the list of expected endpoints using `HasEndpoint`, which uses the `Equal` method for endpoints. If this is not the case when the timer runs out, `ElementsMatch` is called to get better logging about the differences. This is the same reason that `ElementsMatch` is called in the code that I wrote here.

Reply:
But then we have a situation here where the `Equal` method returns false, but `ElementsMatch` considers those elements to be equal, so we neither get the desired result in the test, nor the additional logging that we require. We may need to write our own which lets us use a custom Equals function - something like:

```go
func ElementsMatchFunc[N](t *testing.T, expected []N, actual []N, equal func(a, b N) bool) bool {
    // ... 
} 
```

The main difficulty I think will be constructing the appropriate logging message for the failure but we can probably use [go-spew](https://github.com/davecgh/go-spew) or [litter](https://github.com/sanity-io/litter) to help with this